### ISSUE TITLE
Shared JIT Code Cache + Aggregated Profiling for `interp.Pool`

### ISSUE THREAD

**Issue by siyul-park** [siyul-park/minivm#59]
minivm's `interp.Pool` creates up to N interpreters from one `program.Program`. Today each member independently profiles, hits the JIT threshold, compiles, and mmaps the same hot functions: N× compile cost, N× executable memory, N× slower warmup. The roadmap explicitly lists "Share JIT code cache and aggregate profiling across `interp.Pool` interpreters" as the fix. This plan adds a shared `Cache` that (a) aggregates tick counts so the pool reaches the threshold ~N× faster, (b) compiles each hot function exactly once (single-flight), (c) lets every member install the shared native code into its own closure table, and (d) exposes an aggregated `Pool.Profile()`.

Bonus fix: `WithProfile(p)` passed to `NewPool` today makes N goroutines call non-thread-safe `prof.Stats.Add` concurrently — already racy. Document it; the new aggregated profile is the safe path.

## Why sharing is safe (verified in code)

- Native code receives all interpreter state per call via the scratch argv (`scratch()`, interp.go:842): stack base, globals base, BP, SP, journal pointer. No per-interpreter pointers baked into emitted code.
- Compile-time-baked values (constant immediates, direct-call target heap addrs in `call()`, jit_arm64.go:534) are deterministic per Program — `New` allocates heap constants in program order, so all pool members have identical layouts. `asm.Callable`s are valid in any member.
- Globals-kind assumptions are guarded at runtime (`guardRef`/`guardI64` deopt exits) — cross-member divergence deopts safely.
- **Hard blocker handled**: `asm.Buffer.Write` flips the active mapping RW (`unseal`, asm/buffer.go:150) during writes. Sharing one growable buffer would SIGBUS members executing from it mid-compile. Solution: one **sealed-once buffer per publication**, never written after publish.

### New file `interp/cache.go`

```go
type Cache struct {
    stats   *prof.Stats                    // aggregated profile, merged cold-path under mu
    modules atomic.Pointer[[]*jitModule]   // COW append-only publication list
    buffers []*asm.Buffer                  // sealed code regions, freed at last release
    samples []atomic.Int64                 // per-func aggregated tick counts
    states  []atomic.Int32                 // per-func: cold | building | done
    refs    atomic.Int64                   // owner + attached Interpreters
    mu      sync.Mutex                     // serializes compile/publish + stats merge
}

func NewCache(prog *program.Program) *Cache   // sizes by len(prog.Constants)+1; refs=1
func (c *Cache) Profile() prof.Snapshot
func (c *Cache) Close() error                 // drop owner ref; free buffers at 0
```

Private: `attach`/`detach` (refcount, free at zero), `due(addr, threshold)` (trigger), `publish(mod, buf)`, `fail(addr)`, `flush(s *prof.Stats)` (merge+reset under mu). File follows §2.4 slot order.

### Concurrency model

- **Trigger (single-flight)**: in `due()` — `states[addr].Load() != cold → false`; else `samples[addr].Add(1) >= threshold && states[addr].CompareAndSwap(cold, building)`. Exactly one member wins and compiles on its own goroutine at its own safepoint. Leading `Load` gate stops cache-line ping-pong once compiled. `>=` not `==` (concurrent adders can step past). Per-IP histograms stay in each member's local uncontended `prof.Stats`; segment selection uses the winner's local histogram (~threshold/N samples, proportionally representative).
- **Publication**: COW slice behind `atomic.Pointer` + per-member generation index `jitGen int`. Each safepoint: one atomic load; if `len(*mods) > i.jitGen`, install tail via existing `i.install`, advance. Wait-free hot path; members created later (lazy `grow`) start at gen 0 and catch up on first safepoint.
- **Compilation serialized under `c.mu`**: `complete()` compiles whole call components incl. callees — concurrent overlapping compiles would double-publish; serialization is free (compiles are rare).
- **Code memory**: winner builds throwaway `jitCompiler` (fresh buffer), compiles, links, then transfers the buffer to `cache.buffers`. Never written again → no W^X flip under concurrent execution. Solo interpreters keep today's reused private buffer.
- **Closure binding stays per-interpreter**: only `Callable`s shared. Each member wraps in own `i.entry`/`i.segment` closures and patches own `i.code`/`i.fallbacks` on its own goroutine — journal stays per-interpreter (pointer is per-call anyway).
- **Install idempotency guard**: in `install`, skip entry replacement when `i.fallbacks[addr] != nil`. Prevents double-install recording a JIT wrapper as fallback → live-lock via `entry()` trap path (interp.go:769). Also hardens solo path.

### `interp/interp.go` changes

- Fields: `cache *Cache` (lifecycle layer), `jitGen int` (mutable counters, near `fp`/`sp`) per §2.5. New `WithCache(c *Cache)` option.
- `New`: `cache.attach()` when set. `Close`: `cache.detach()` instead of freeing private compiler (cache mode: `i.compiler` stays nil).
- `safepoint()` (after local `i.prof.Add`, interp.go:683):
  - cache mode: `if i.threshold >= 0 && i.cache.due(f.addr, i.threshold) { i.jit(f.addr) }` then `i.sync()`.
  - solo mode: today's `==` check untouched.
- `jit(addr)` cache branch: throwaway compiler, `Compile`; success → `cache.publish(mod, buf)`; failure → `cache.fail(addr)` + `JITAdd(Errors:1)`, propagate. Winner installs via same `sync()` path as consumers (symmetry, §0.4).
- New private `sync()`: load modules, install `(*mods)[i.jitGen:]`, advance.
- Private `flush()` bridge: `i.cache.flush(i.prof)` (keeps `prof` unexported, §1.5).
- Stub arch: `newJITCompiler` nil → publish nothing, mark done. Identical behavior on !arm64.
- **No `WithCache` → zero behavior change** (every branch gated on `i.cache != nil`).

### `interp/pool.go` changes

- `NewPool`: `cache := NewCache(prog)`; append `WithCache(cache)` next to `WithMarshaler(&codec{})` (pool.go:37); store on `Pool`.
- `Put` + `drop`: flush member's local profile via `i.flush()` (merge-then-reset; repeated cycles never double-count).
- `Close`: after draining idle, `p.cache.Close()`. Refcount means executable pages freed only when last outstanding member drops — fixes would-be use-after-free of native code after `Pool.Close`.
- New public `(p *Pool) Profile() prof.Snapshot` → `p.cache.Profile()` (documented "as of last Put").

### `prof/prof.go` changes

- `(p *Stats) Merge(s Snapshot)`, `(p *Stats) Reset()` — plain methods, Cache guards with mu. Document `WithProfile`-into-`NewPool` race in `WithProfile` doc comment.

## Hazards → resolutions

| Hazard | Resolution |
|---|---|
| W^X flip on shared buffer during execution | sealed-once buffer per publication |
| Per-interpreter closure tables | share Callables only; members install on own goroutine at safepoint |
| Deopt journal per-interpreter | unchanged — journal pointer is per-call scratch arg |
| Duplicate component compile (`complete()` covers callees) | compiles under mu; publish CASes covered addrs done; install skips shadowed |
| Double install → live-lock in trap path | `fallbacks[addr] != nil` guard in `install` |
| Profile contention | local Stats on hot path; 1 atomic add/tick for trigger, gated once done; merge only at Put/Close |
| Exec memory freed while member runs native | refcounted release |
| Members created after publication | gen index 0 → catch-up at first safepoint |
| `==` trigger race | `>=` + CAS |
| !arm64 | stub → mark done, behavior identical |

## Build order

1. `prof/prof.go`: `Merge`, `Reset` + `TestStats_Merge`/`TestStats_Reset` in `prof/prof_test.go`.
2. `interp/cache.go` (new) + `interp/cache_test.go`: `TestNewCache`, `TestCache_Profile`, `TestCache_Close` (public symbols only, §6.3; exercise due/publish through Interpreter/Pool behavior).
3. `interp/interp.go`: option, fields, safepoint branch, jit cache path, `sync`, install guard, Close detach, flush bridge. `TestWithCache` in `interp_test.go` (reuse `requireJIT(t)` skip pattern, interp_test.go:2452).
4. `interp/pool.go`: cache wiring, Put/drop flush, Close release, `Pool.Profile`. Extend `pool_test.go`.
5. Docs: `docs/roadmap.md` (line 32 item), `docs/architecture.md`, `docs/jit-internals.md` (sharing/lifetime contract: published buffers immutable; Callables shared, bindings per-interpreter).

## Verification

- `go build ./... && go vet ./...`
- `go test -race ./interp/... ./prof/...` — full suite; existing tests must pass untouched (solo behavior unchanged).
- Behaviors:
  - **Single compile across N**: pool N=4, hot loop concurrent on all members, `WithTick(1), WithCutoff(1), WithThreshold(N)` (threshold reachable only by aggregation) → `p.Profile().JIT.Attempts == 1`, results match threaded-only run.
  - **Consumer installs without compiling**: member borrowed after compile returns correct results, aggregate Attempts stays 1.
  - **Concurrency correctness**: N goroutines × M Get/Run/Put cycles, jittable + deopt paths, results equal threaded reference, `-race` clean.
  - **Lifetime**: `Pool.Close` with outstanding member mid-native-Run → finishes, Put succeeds, no crash; idempotent Close; manual shared `Cache.Close` frees only after last `Interpreter.Close`.
  - **Profile aggregation**: `Pool.Profile()` equals sum of member work; repeated Puts don't double-count.
  - **Trigger aggregation**: threshold T, N members × T/N samples each → fires; single member with T/N → doesn't.
- On arm64 darwin: run `make bench` (or fib benchmark) with pool to confirm warmup improvement and no regression.

**Comment by siyul-park** [comment]
Triage update: keep open, but the remaining scope is narrower than the original issue.

Resolved/mostly implemented in current `main`:
- `interp/cache.go` now has a shared `Cache` with per-function state, sealed executable buffers, `attach`/`detach`, `due`, `publish`, and refcounted lifetime.
- `interp/pool.go` creates one cache per pool, passes `WithCache(cache)` into members, flushes on `Put`/`drop`, and closes the cache on `Pool.Close`.

Still unresolved before this issue should close:
- The original issue asks for an aggregated public profile API such as `Pool.Profile()` / `Cache.Profile()`, but `Pool` currently has no public `Profile` method and `Cache` does not own profile aggregation.
- Decide whether the newer shared `prof.Profiler` path supersedes the originally proposed `Cache.Profile()` design. If yes, update this issue’s acceptance criteria and docs to point to that API instead of adding another profile surface.
- Add/confirm explicit tests for single compile across multiple pool members, consumer install without recompiling, outstanding-member lifetime after `Pool.Close`, and shared-profile behavior.

Suggested completion path:
1. Either implement `Pool.Profile()` or rewrite the issue to say pooled users should pass a shared `prof.Profiler` via `WithProfiler`.
2. Add focused `interp` tests covering shared compile/lifetime/profile behavior.
3. Update `docs/profile.md` / `docs/architecture.md` with the final pool profiling contract.
